TestAlertmanagerSharding is flaky due to a logic issue

**Describe the bug**
The `TestAlertmanagerSharding`, which was updated in #3839, is flaky. As an example, you can see it [here](https://github.com/cortexproject/cortex/runs/2054919160) and [here](https://github.com/cortexproject/cortex/actions/runs/631672519).

**To Reproduce**
've reproduced it locally with debug logs. This is a snippet of logs:

```
# we create the silence here

17:57:39 alertmanager-3: level=debug ts=2021-03-08T16:57:39.9726694Z caller=multitenant.go:929 component=MultiTenantAlertmanager msg="user does not have an alertmanager in this instance" user=user-5
17:57:39 alertmanager-1: level=debug ts=2021-03-08T16:57:39.9740552Z caller=multitenant.go:907 component=MultiTenantAlertmanager msg="user not found while trying to replicate state" user=user-5 key=sil:user-5

# the periodic sync realise the shard has changed and loads user-5 too, but it's "too late"

17:57:40 alertmanager-3: level=debug ts=2021-03-08T16:57:40.1726301Z caller=multitenant.go:673 component=MultiTenantAlertmanager msg="setting config" user=user-5
17:57:40 alertmanager-3: level=debug ts=2021-03-08T16:57:40.1730083Z caller=multitenant.go:726 component=MultiTenantAlertmanager msg="initializing new per-tenant alertmanager" user=user-5
17:57:40 alertmanager-3: level=debug ts=2021-03-08T16:57:40.1732274Z caller=alertmanager.go:139 user=user-5 msg="starting tenant alertmanager with ring-based replication"
```

The problem is that if the silence is created soon after a resharding, the replication may fail.

**Expected behavior**
The replication should not fail if happening right after a resharding.

`TestAlertmanagerSharding` is still flaky. I've got a CI run failing with this output:

```
2021-06-22T15:01:42.5842285Z === RUN   TestAlertmanagerSharding/RF_=_3
2021-06-22T15:01:42.5845071Z 14:41:44 Starting consul
2021-06-22T15:01:42.5845509Z 14:41:45 consul: ==> Starting Consul agent...
2021-06-22T15:01:42.5846185Z 14:41:45 consul: Version: '1.8.4'
2021-06-22T15:01:42.5851377Z 14:41:45 consul: Node ID: 'aece4fca-c8f9-0bf9-a4bf-34cdc2fdc092'
2021-06-22T15:01:42.5852201Z 14:41:45 consul: Node name: 'consul'
2021-06-22T15:01:42.5852825Z 14:41:45 consul: Datacenter: 'dc1' (Segment: '<all>')
2021-06-22T15:01:42.5853375Z 14:41:45 consul: Server: true (Bootstrap: false)
2021-06-22T15:01:42.5854121Z 14:41:45 consul: Client Addr: [0.0.0.0] (HTTP: 8500, HTTPS: -1, gRPC: 8502, DNS: 8600)
2021-06-22T15:01:42.5854659Z 14:41:45 consul: Cluster Addr: 127.0.0.1 (LAN: 8301, WAN: 8302)
2021-06-22T15:01:42.5855599Z 14:41:45 consul: Encrypt: Gossip: false, TLS-Outgoing: false, TLS-Incoming: false, Auto-Encrypt-TLS: false
2021-06-22T15:01:42.5856380Z 14:41:45 consul: ==> Log data will now stream in as it occurs:
2021-06-22T15:01:42.5856879Z 14:41:45 consul: ==> Consul agent running!
2021-06-22T15:01:42.5860826Z 14:41:45 Ports for container: e2e-cortex-test-consul Mapping: map[8500:49233]
2021-06-22T15:01:42.5861625Z 14:41:45 Starting minio-9000
2021-06-22T15:01:42.5862456Z 14:41:46 minio-9000: Attempting encryption of all config, IAM users and policies on MinIO backend
2021-06-22T15:01:42.5863529Z 14:41:46 Ports for container: e2e-cortex-test-minio-9000 Mapping: map[9000:49235]
2021-06-22T15:01:42.5864356Z 14:41:47 Starting alertmanager-1
2021-06-22T15:01:42.5865490Z 14:41:48 alertmanager-1: level=warn ts=2021-06-22T14:41:48.208554093Z caller=experimental.go:19 msg="experimental feature in use" feature="Alertmanager sharding"
2021-06-22T15:01:42.5866854Z 14:41:48 Ports for container: e2e-cortex-test-alertmanager-1 Mapping: map[80:49241 9094:49239 9095:49237]
2021-06-22T15:01:42.5867730Z 14:41:49 Starting alertmanager-2
2021-06-22T15:01:42.5868911Z 14:41:50 alertmanager-2: level=warn ts=2021-06-22T14:41:50.05556098Z caller=experimental.go:19 msg="experimental feature in use" feature="Alertmanager sharding"
2021-06-22T15:01:42.5870286Z 14:41:50 Ports for container: e2e-cortex-test-alertmanager-2 Mapping: map[80:49247 9094:49245 9095:49243]
2021-06-22T15:01:42.5871334Z 14:41:52 Starting alertmanager-3
2021-06-22T15:01:42.5872637Z 14:41:52 alertmanager-3: level=warn ts=2021-06-22T14:41:52.861184798Z caller=experimental.go:19 msg="experimental feature in use" feature="Alertmanager sharding"
2021-06-22T15:01:42.5873995Z 14:41:52 Ports for container: e2e-cortex-test-alertmanager-3 Mapping: map[80:49253 9094:49251 9095:49249]
2021-06-22T15:01:42.5874756Z     alertmanager_test.go:544: 
2021-06-22T15:01:42.5875254Z         	Error Trace:	alertmanager_test.go:544
2021-06-22T15:01:42.5875759Z         	Error:      	elements differ
2021-06-22T15:01:42.5876137Z         	            	
2021-06-22T15:01:42.5876516Z         	            	extra elements in list A:
2021-06-22T15:01:42.5876968Z         	            	([]interface {}) (len=1) {
2021-06-22T15:01:42.5877385Z         	            	 (string) (len=7) "alert_3"
2021-06-22T15:01:42.5877822Z         	            	}
2021-06-22T15:01:42.5878115Z         	            	
2021-06-22T15:01:42.5878414Z         	            	
2021-06-22T15:01:42.5878727Z         	            	listA:
2021-06-22T15:01:42.5879093Z         	            	([]string) (len=3) {
2021-06-22T15:01:42.5879486Z         	            	 (string) (len=7) "alert_1",
2021-06-22T15:01:42.5879912Z         	            	 (string) (len=7) "alert_2",
2021-06-22T15:01:42.5880321Z         	            	 (string) (len=7) "alert_3"
2021-06-22T15:01:42.5880685Z         	            	}
2021-06-22T15:01:42.5880985Z         	            	
2021-06-22T15:01:42.5881276Z         	            	
2021-06-22T15:01:42.5881592Z         	            	listB:
2021-06-22T15:01:42.5881941Z         	            	([]string) (len=2) {
2021-06-22T15:01:42.5882347Z         	            	 (string) (len=7) "alert_1",
2021-06-22T15:01:42.5882789Z         	            	 (string) (len=7) "alert_2"
2021-06-22T15:01:42.5883150Z         	            	}
2021-06-22T15:01:42.5883660Z         	Test:       	TestAlertmanagerSharding/RF_=_3
2021-06-22T15:01:42.5884258Z     alertmanager_test.go:544: 
2021-06-22T15:01:42.5884867Z         	Error Trace:	alertmanager_test.go:544
2021-06-22T15:01:42.5885369Z         	Error:      	elements differ
2021-06-22T15:01:42.5885735Z         	            	
2021-06-22T15:01:42.5886125Z         	            	extra elements in list A:
2021-06-22T15:01:42.5886579Z         	            	([]interface {}) (len=1) {
2021-06-22T15:01:42.5886996Z         	            	 (string) (len=7) "alert_3"
2021-06-22T15:01:42.5887360Z         	            	}
2021-06-22T15:01:42.5887652Z         	            	
2021-06-22T15:01:42.5887951Z         	            	
2021-06-22T15:01:42.5888263Z         	            	listA:
2021-06-22T15:01:42.5888626Z         	            	([]string) (len=3) {
2021-06-22T15:01:42.5889018Z         	            	 (string) (len=7) "alert_1",
2021-06-22T15:01:42.5889440Z         	            	 (string) (len=7) "alert_2",
2021-06-22T15:01:42.5889854Z         	            	 (string) (len=7) "alert_3"
2021-06-22T15:01:42.5890213Z         	            	}
2021-06-22T15:01:42.5890516Z         	            	
2021-06-22T15:01:42.5890807Z         	            	
2021-06-22T15:01:42.5891126Z         	            	listB:
2021-06-22T15:01:42.5891476Z         	            	([]string) (len=2) {
2021-06-22T15:01:42.5891884Z         	            	 (string) (len=7) "alert_1",
2021-06-22T15:01:42.5892293Z         	            	 (string) (len=7) "alert_2"
2021-06-22T15:01:42.5892655Z         	            	}
2021-06-22T15:01:42.5893161Z         	Test:       	TestAlertmanagerSharding/RF_=_3
2021-06-22T15:01:42.5893936Z 14:41:56 Killing alertmanager-3
2021-06-22T15:01:42.5894530Z 14:41:56 Killing alertmanager-2
2021-06-22T15:01:42.5895130Z 14:41:56 Killing alertmanager-1
2021-06-22T15:01:42.5895680Z 14:41:56 Killing minio-9000
2021-06-22T15:01:42.5896074Z 14:41:57 Killing consul
2021-06-22T15:01:42.5896749Z --- FAIL: TestAlertmanagerSharding (24.97s)
2021-06-22T15:01:42.5899663Z     --- PASS: TestAlertmanagerSharding/RF_=_2 (11.99s)
2021-06-22T15:01:42.5900559Z     --- FAIL: TestAlertmanagerSharding/RF_=_3 (12.98s)
```

--- a/integration/alertmanager_test.go
+++ b/integration/alertmanager_test.go
@@ -531,6 +531,12 @@ func TestAlertmanagerSharding(t *testing.T) {
 				require.NoError(t, err)
 				err = c3.SendAlertToAlermanager(context.Background(), alert(3, 2))
 				require.NoError(t, err)
+
+				// Wait for the alerts to be received by every replica.
+				require.NoError(t, alertmanagers.WaitSumMetricsWithOptions(
+					e2e.Equals(float64(3*testCfg.replicationFactor)),
+					[]string{"cortex_alertmanager_alerts_received_total"},
+					e2e.SkipMissingMetrics))
 			}
 
 			// Endpoint: GET /v1/alerts

